picamera2/controls.py

```diff
@@ -0,0 +1,160 @@
+        the camera start function. Users can also use it check to see what
+        they defined the controls as BEFORE they start the camera.
+
+        To check the controls after the camera has started, please use the
+        current function.
+
+        @return
+            Returned is a dictionary of camera controls that the user wants to
+            change. If all controls are set to none, then the returned dict
+            is empty, which is still accepted by the camera_manager.
+        """
+        controls = {}
+        for attr, types in self.__annotations__.items():
+            val = getattr(self, attr)
+            if (not isinstance(val, types) and
+                    val is not None and
+                    attr not in str(val)):
+                raise TypeError(f"{attr} must be of these type(s): {types}")
+            if isinstance(val, str):  # Assume val is mode if a string.
+                val = getattr(getattr(libcamera, attr), val)
+            controls[attr] = val
+        user_defined = {k: v for k, v in controls.items() if v is not None}
+        user_defined = dict(sorted(user_defined.items()))
+        return user_defined
+
+    @property
+    def current(self) -> dict:
+        """
+        This function returns the current user-defined settings of the camera.
+        It is only updated through the CompletedRequests module.
+        """
+        return self._current
+
+    @property
+    def _blank(self) -> None:
+        """
+        This function _blanks all controls to None.
+        Its only use is to clear the controls at camera close. This is so
+        that if a user decides to reopen the camera after close, the old
+        control settings do not persist and carry over to the next start.
+        """
+        for attr in self.__annotations__.keys():
+            setattr(self, attr, None)
+
+    @property
+    def fps(self) -> int:
+        return self._fps
+
+    @fps.setter
+    def fps(self, value) -> None:
+        """
+        Converts a frames per seconds (fps) value to a microseconds value that
+        the camera accepts.
+
+        @param value
+            A value with units of frames per second.
+        @return
+            A tuple of values indicated a microseconds value that enables
+            a fixed number of frames per second.
+
+        e.g. In: fps = 30
+             Out: FrameDurationLimits = (33333,33333)
+        """
+        uS = int(1e6 / value)
+        self.FrameDurationLimits = (uS, uS)
+
+    @property
+    def ranges(self) -> dict:
+        """This function returns the possible ranges for each controls."""
+        _ranges = {}
+        for ctrl, val_tuple in self._camera_controls.items():
+            _ranges[ctrl] = {'min': val_tuple[0], 'max': val_tuple[1]}
+        _ranges = dict(sorted(_ranges.items()))
+        return _ranges
+
+    @property
+    def defaults(self) -> dict:
+        """This function takes the default camera controls and displays them
```

Comment:
This one makes me a bit nervous. I'm worried that users might think that `'picam2.set_controls(picam2.controls.defaults))` is a good idea... it resets all the camera controls to its normal defaults, right? Only it really isn't!

Reply:
Would the proper method to reset to defaults then be to stop, close and then restart?

Reply:
I agree. It was a hopeful addition.